napari/utils/actions/_registries.py

```diff
@@ -0,0 +1,326 @@
+from __future__ import annotations
+
+from concurrent.futures import Future, ThreadPoolExecutor
+from typing import TYPE_CHECKING, Callable, Optional
+
+from psygnal import Signal
+
+from ...utils.translations import trans
+from ._menus import MenuGroup, MenuId
+from ._types import (
+    MenuItem,
+    SubmenuItem,
+    _RegisteredCommand,
+    _RegisteredKeyBinding,
+)
+from ._util import MockFuture
+
+if TYPE_CHECKING:
+    from typing import Dict, Iterator, List, Sequence, Set, Tuple, Union
+
+    from ._types import (
+        CommandHandler,
+        CommandId,
+        CommandRule,
+        KeybindingRule,
+        TranslationOrStr,
+    )
+
+    DisposeCallable = Callable[[], None]
+    CommandDecorator = Callable[[CommandHandler], CommandHandler]
+    MenuOrSubmenu = Union[MenuItem, SubmenuItem]
+
+
+class CommandsRegistry:
+
+    registered = Signal(str)
```

Comment:
psygnal actually needs a real type, and you can't do `isinstance(something, NewType())`

```python
In [1]: isinstance('asdf', NewType('S', str))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [1], in <cell line: 1>()
----> 1 isinstance('asdf', NewType('S', str))

TypeError: isinstance() arg 2 must be a type or tuple of types
```

Reply:
but NewType assign `__supertype__` property of type definition. So it is possible to check if `t.__qualname__ == 'NewType.<locals>.new_type'` the check could be changed from `isinstance(val, t)` to `isinstnace(val, t.__supertype__`)

Reply:
that would need to be done in psygnal

Reply:
for me this reduce confussion. But from reading more about NewType i do not understand why 
`CommandId = NewType('CommandId', str)` is better than `CommandId = str`. 
When using the current approach the type checker will comply with that value is `str` not `CommandId`

Reply:
the main reason is if you've got a number of different strings floating around, they're all just strings, but you don't want to confuse them.  So, for example, put this into your IDE:

```python
from typing import NewType
from pydantic import BaseModel

MenuId = NewType('MenuId', str)
CommandId = NewType('CommandId', str)


def get_menu(id: MenuId):
    ...


def get_command(id: CommandId):
    ...


class MenuRule(BaseModel):
    id: MenuId


class CommandRule(BaseModel):
    id: CommandId


menu_rule = MenuRule(id='/my/menu')
command_rule = CommandRule(id='some.command')


get_menu(command_rule.id)
```

As is: that will give a type error, because we're confusing two different types of strings in the `get_menu` call.  If you used Type Aliases (`CommandId = str`) you won't get that.  So, with a number of mutually incompatible `str` types flying around, it's nice to have.